app/models/user.rb

```diff
@@ -37,6 +37,28 @@ def possible_score(competition)
   end
 
   def matches(competition: nil)
+    cached_results_array = Rails.cache.read("#{cache_key_with_version}-matches")&.to_a
+    if run_matches_query?(cached_results_array) 
+      query_results_array = execute_matches_query(competition: competition).to_a
+      Rails.cache.write("#{cache_key_with_version}-matches", query_results_array)
+      return query_results_array
+    else
+      return cached_results_array
+    end
+  end
+
+  private
+
+  def run_matches_query?(cached_results_array)
+    cached_results_array.nil? || matches_cached_result_expired?(cached_results_array)
+  end
+
+  def matches_cached_result_expired?(cached_results_array)
+    next_upcoming_match_according_to_cache = cached_results_array.select { |match| match['status'] === 'upcoming' }.map { |match| match['kickoff_time']}.sort.first
+    next_upcoming_match_according_to_cache < DateTime.now
+  end
```

Comment:
This seems like it wouldn't expire the cache when scores are updated for a match that is already `started`, no?

We could instead maybe simply compare the latest `updated_at` timestamp from the cached results with the latest `updated_at` timestamp from the matches DB table? It does require a query on the DB, but it should be a pretty light one, so we would still benefit from caching. For example:

```suggestion
  def matches_cached_result_expired?(cached_results_array, competition)
    matches_last_updated_at_according_to_cache = cached_results_array.max { |match| match['updated_at'] }['updated_at']
    matches_last_updated_at_according_to_cache > (competition&.matches || Match).maximum(:updated_at)
  end
```

Reply:
@trouni Thanks for the review. I just updated the PR after this to refactor the score method to also include using cache in a similar way.

I think I see the problem, that with my method above the cache would only get expired once and not again after the end of the game. Is that right?

I wasn't sure how often the match table gets updated. Is it updated at the end of every match? If not another method might be just to do as above but add 120 minutes to DateTime.now so that the match has finished.

Reply:
Also @trouni it might not be working for you because of the line ` config.action_controller.perform_caching = false` in `development.rb`.
I didn't realise this line was in there on the earlier commit but you may need to change to true.

Reply:
> I think I see the problem, that with my method above the cache would only get expired once and not again after the end of the game. Is that right?

Yes, that's the idea. Unfortunately, we don't know exactly when the matches get updated with the scores since it depends on the 3rd party API service that we use. We're fetching the API every 10 minutes, but results are sometimes updated more than an hour after the end of the game (there's issue #97 open about that problem).

Also, the knockout rounds will last longer than the group matches, and ideally the app should be agnostic to the sport that is being played (so that it could be used for rugby or other competitions for example). That's why I think it's best to rely on whether the matches were actually updated rather than hardcoding a specific time when to expire the cache.

Reply:
Another thing: in the ideal scenario, we would have live match scores in our app as well (the ideal outcome of #97), which would require expiring the cache much more often.